Fix deadlock in `txsub.System.Tick()` and `tx_bad_seq` errors

bartekn opened the pull request:
This commit fixes two issues in `txsub` package. First, issue was a
deadlock in `Tick()` function:

```go
sys.tickMutex.Lock()
if sys.tickInProgress {
  logger.Debug("ticking in progress")
  return
}
sys.tickInProgress = true
sys.tickMutex.Unlock()
```

When two `Tick()` methods were called simultanously one of them could
`Lock()` the mutex without calling `Unlock()` properly. This could
basically break the `txsub` system as sequence number could not be
updated.

The second issue was a problem with `App` initialization.
`App.networkPassphrase` was set to test network passphrase by default.
If `txsub` system was initialized before updating the network passphrase
using connected `stellar-core`, the system would not be able to
correctly find the transaction result in a database as calculated hash
was different than the actual hash of the transaction.

Close #305.

tomquisel commented on services/horizon/internal/app.go:
```diff
@@ -270,6 +270,13 @@ func (a *App) UpdateStellarCoreInfo() {
+	}
+
 	a.coreVersion = resp.Info.Build
 	a.networkPassphrase = resp.Info.Network
```
This line can be removed now, right?

tomquisel commented on services/horizon/internal/app.go:
```diff
@@ -270,6 +270,13 @@ func (a *App) UpdateStellarCoreInfo() {
+	// Check if NetworkPassphrase is different, if so exit Horizon as it can break the
+	// state of the application.
+	if resp.Info.Network != a.networkPassphrase {
+		log.Error("Network passhprase of stellar-core does not match Horizon configuration. Exiting...")
```
change `passhprase` to `passphrase`, and maybe improve error message to print the two passphrase values.

tomquisel commented on services/horizon/internal/app.go:
```diff
@@ -64,7 +64,7 @@ func NewApp(config Config) (*App, error) {
 	result := &App{config: config}
 	result.horizonVersion = app.Version()
-	result.networkPassphrase = build.TestNetwork.Passphrase
+	result.networkPassphrase = config.NetworkPassphrase
```
Could we remove `App.networkPassphrase` and just use `App.Config.NetworkPassphrase` as the single source of truth?

tomquisel commented on services/horizon/internal/txsub/system.go:
```diff
@@ -178,7 +187,8 @@ func (sys *System) Tick(ctx context.Context) {
 	// Make sure Tick is not run concurrently
 	sys.tickMutex.Lock()
```
I might accomplish this locking with two functions:

`SetTickInProgress`:
```
        sys.tickMutex.Lock()
        defer sys.tickMutex.Unlock()
	if sys.tickInProgress {
		logger.Debug("ticking in progress")
		return false
	}
	sys.tickInProgress = true
        return true
```

`UnsetTickInProgress`:
```
        sys.tickMutex.Lock()
        defer sys.tickMutex.Unlock()
        sys.tickInProgress = false 
```

and then call those from `Tick` (which checks the results of `SetTickInProgress` and exits if `false`, and calls `UnsetTickInProgress()` in `defer`.

I think that'd make this a little safer from this kind of bug in the future.

bartekn replied:
Good point. The ultimate solution is to remove global ticking service completely (#613). It caused a lot of concurrency problems in the past (including this one) and can be implemented much easier using infinite loop.

tomquisel commented on services/horizon/main.go:
```diff
@@ -239,6 +239,10 @@ func initConfig() {
+	if viper.GetString("network-passphrase") == "" {
+		stdLog.Fatal("Invalid config: network-passphrase is blank.  Please specify --network-passphrase on the command line or set the NETWORK_PASSPHRASE environment variable.")
+	}
+
```
Hrm, is it a breaking change to make this required? It seems like it might be, and it'd be better to leave `network-passphrase` optional + default to testnet as we do currently.

bartekn replied:
Good point, set default value to testnet passphrase but in the next minor we need to make this empty by default. Ex. if you run ingestion node without web server and when stellar-core is not running it should be explicitly set by Horizon admin (#817).

oryband commented:
I've tested this branch in an environment where I can reproduce the issue and the problem is solved.

bartekn commented:
@oryband thanks for checking this. Would you mind checking this again in your environment (I pushed one last commit with updates).
@tomquisel all good points, fixed and added 2 comments.

oryband commented:
@bartekn I tested this again. It works. Thanks for sorting this issue out! :trophy:

oryband commented:
@bartekn do you know when you'll release a new stable version that includes this branch?

bartekn commented:
@oryband today.

oryband commented:
WHOOHOO.

howardtw commented on services/horizon/internal/txsub/system.go:
```diff
@@ -170,25 +179,40 @@ func (sys *System) submitOnce(ctx context.Context, env string) SubmissionResult
 	return sr
 }
 
+// setTickInProgress sets `tickInProgress` to `true` if it's not
```
Looks like we set `tickInProgress` to `true` only if it is `false`?

bartekn replied:
Good catch! https://github.com/stellar/go/pull/824

howardtw commented on services/horizon/internal/txsub/system.go:
```diff
@@ -73,12 +74,20 @@ func (sys *System) Submit(ctx context.Context, env string) (result <-chan Result
 	r := sys.Results.ResultByHash(ctx, info.Hash)
 
-	if r.Err != ErrNoResults {
+	if r.Err == nil {
```
Nothing wrong with the logic here but what is the reason that we have an Err field in r as oppose to have ResultByHash return r and err directly?

bartekn replied:
It's also used as an object returned by channel.
